modules/repository-datastax/src/main/java/com/intuit/wasabi/repository/cassandra/impl/CassandraAssignmentsRepository.java

```diff
@@ -183,6 +203,146 @@ public CassandraAssignmentsRepository(
         return result;
     }
 
+    /**
+     * Populate experiment metadata asynchronously at one place.
+     *
+     * experimentIds is NULL when called from AssignmentsResource.getBatchAssignments() => api:/v1/assignments/applications/{applicationName}/users/{userID}
+     * experimentBatch.labels are NULL when called from AssignmentsImpl.doPageAssignments() => api:/v1/assignments/applications/{applicationName}/pages/{pageName}/users/{userID}
+     *
+     * @param userID Input: Given user id
+     * @param appName Input: Given application name
+     * @param context Input: Given context
+     * @param experimentBatch Input/Output: Given experiment batch. This object will be modified and become one of the output; in the case of AssignmentsImpl.doPageAssignments()
+     * @param experimentIds Input: Given experiment ids.
+     * @param prioritizedExperimentList Output: prioritized experiment list of ALL the experiments for the given application.
+     * @param experimentMap Output: Map of 'experiment id TO experiment' of ALL the experiments for the given application.
+     * @param existingUserAssignments Output: ALL the existing user assignments of given user_id, application name and context.
+     * @param bucketMap Output: Map of 'experiment id TO BucketList' of ONLY experiments which are associated to the given application and page.
+     * @param exclusionMap Output: Map of 'experiment id TO to its mutual experiment ids' of ONLY experiments which are associated to the given application and page.
+     *
+     */
+    @Override
+    @Timed
+    public void populateExperimentMetadata( User.ID userID, Application.Name appName, Context context, ExperimentBatch experimentBatch, Set<Experiment.ID> experimentIds,
+                                            PrioritizedExperimentList prioritizedExperimentList,
+                                            Map<Experiment.ID, com.intuit.wasabi.experimentobjects.Experiment> experimentMap,
+                                            Table<Experiment.ID, Experiment.Label, String> existingUserAssignments,
+                                            Map<Experiment.ID, BucketList> bucketMap,
+                                            Map<Experiment.ID, List<Experiment.ID>> exclusionMap
+                                            ) {
+        if(logger.isDebugEnabled()) logger.debug("populateExperimentMetadata - STARTED: userID={}, appName={}, context={}, experimentBatch={}, experimentIds={}", userID, appName, context, experimentBatch, experimentIds);
+        if(userID==null || appName==null || context==null || experimentBatch==null || (experimentBatch.getLabels()==null && (experimentIds==null || experimentIds.isEmpty())) ) {
```

Comment:
Removed nulls as it is already taken care of in previous methods and added Optional implementation for allowAssignments where NULL/empty is expected in some cases...